Remove Region and add new Clip Extension

JimBobSquarePants opened the pull request:
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp.Drawing/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
<!-- A description of the changes proposed in the pull-request -->
Fixes #105 

- Removes all references to `Region` and inheriting types.
- Adds new `Clip(IImageProcessingContext, IPath, Action<IImageProcessingContext>)` extension with functionality copied from [samples](https://github.com/SixLabors/Samples/blob/38faba5caf866c0168ea67a5e7c37c06d3043775/ImageSharp/CustomImageProcessor/Program.cs)

<!-- Thanks for contributing to ImageSharp.Drawing! -->

JimBobSquarePants commented on src/ImageSharp.Drawing/Processing/Processors/Text/DrawTextProcessor{TPixel}.cs:
```diff
@@ -325,6 +325,11 @@ public void EndGlyph()
                 {
                     IPath path = this.builder.Build();
 
+                    if (path.Bounds.Equals(RectangleF.Empty))
```
I discovered that this was throwing in `TessellatedMultipolygon.Create` during testing with the current fonts library.

JimBobSquarePants replied:
I moved this fix to another PR and added tests

tocsoft commented on src/ImageSharp.Drawing/Processing/Extensions/FillPathExtensions.cs:
```diff
@@ -7,42 +7,30 @@
+            source.Fill(new SolidBrush(color), path);
 
         /// <summary>
         /// Flood fills the image in the shape of the provided polygon with the specified brush..
```
```suggestion
        /// Flood fills the image in the shape of the provided polygon with the specified brush.
```

tocsoft commented on src/ImageSharp.Drawing/Processing/Extensions/FillPathExtensions.cs:
```diff
@@ -7,42 +7,30 @@
     {
         /// <summary>
-        /// Flood fills the image in the shape of the provided polygon with the specified brush.
+        /// Flood fills the image in the shape of the provided polygon with the specified brush..
```
```suggestion
        /// Flood fills the image in the shape of the provided polygon with the specified brush.
```

tocsoft commented on src/ImageSharp.Drawing/Processing/Processors/Drawing/RecursiveImageProcessor.cs:
```diff
@@ -0,0 +1,101 @@
+namespace SixLabors.ImageSharp.Drawing.Processing.Processors.Drawing
+{
+    /// <summary>
+    /// Allows the recursive application of processing operations against and image.
```
```suggestion
    /// Allows the recursive application of processing operations against an image.
```

tocsoft commented on src/ImageSharp.Drawing/Processing/Processors/Drawing/RecursiveImageProcessor.cs:
```diff
@@ -0,0 +1,101 @@
+
+            public void Execute()
+            {
+                // Perform any processing logic you want here, in this case we are going to
```
Comment doesn't read well.... most of it can probably go as the code is relatively self explanatory.... 

I would delete the 3 lines before 
```
// so clone out our source image so we can apply
 // various effects to it without mutating the origional yet.
```

JimBobSquarePants replied:
Done!

tocsoft commented on src/ImageSharp.Drawing/Processing/Processors/Drawing/RecursiveImageProcessor.cs:
```diff
@@ -0,0 +1,101 @@
+                clone.Mutate(x => x.Crop((Rectangle)this.recursiveImageProcessor.Path.Bounds));
+
+                // Use an image brush to apply cloned image as the source for filling the shape
+                var brush = new ImageBrush(clone);
```
We should be able to add an overload of ImageBrush that means we can avoid the crop operation and save memory... it just needs an area of interest of the source/cloned image that its allowed to point with.

i.e. `var brush = new ImageBrush(clone, (Rectangle)this.recursiveImageProcessor.Path.Bounds);`

JimBobSquarePants replied:
I’ll have a look 👍

JimBobSquarePants replied:
Done!

tocsoft commented on src/ImageSharp.Drawing/Processing/Processors/Drawing/RecursiveImageProcessor.cs:
```diff
@@ -0,0 +1,97 @@
+                // Use an image brush to apply cloned image as the source for filling the shape.
+                // We pass explicit bounds to avoid the need to crop the clone;
+                RectangleF bounds = this.recursiveImageProcessor.Path.Bounds;
+                var brush = new ImageBrush(clone, new RectangleF(0, 0, bounds.Width, bounds.Height));
```
0,0 doesn't looks right to me here i would have expected it to have been the top left bound of the Path, so if the shape was Rectangle of 100x100 being applied at 40, 50, then I would have expected we would have to pass in  `(40, 50, 100, 100)`.

tocsoft replied:
I need to try and get my head /debug the changes you made to the Brush to see what's going on but it reads here as a bug as the original code was cropping down to the truncated bounds directly 

https://github.com/SixLabors/Samples/blob/38faba5caf866c0168ea67a5e7c37c06d3043775/ImageSharp/CustomImageProcessor/Program.cs#L102

JimBobSquarePants replied:
That’s exactly how I thought it would work but weirdly it didn’t? That’s why I added those offset tests to be sure.

JimBobSquarePants replied:
@antonfirsov When you're back can you please sense check this. It certainly seems to work but it IS confusing. Thanks!

antonfirsov replied:
Yeah back from vacation since yesterday, I will finally get to this PR this week.

tocsoft replied:
ok I think I've figured out what's wrong with this.. you tried to reuse a region property that was already inside the `ImageBrushApplicator` however that region is specifically supposed to be the target region we well be drawing to not the region of the source image we are cloning from... it happened to work out for you when you did it ion this case due to the fact its ultimately the same source image with and without effects being applied, but ultimately its mixing 2 concerns and it won't work as expected for usages where the image bush is using a source image smaller than the final target where repeating texture output is supposed to be output. I'm looking at exactly what changes are needed inside the brush to make this work correctly and ill create a PR targeting this branch with the required changes.

tocsoft replied:
#154 has the correct fix

JimBobSquarePants replied:
Perfect! Thanks!

antonfirsov commented on src/ImageSharp.Drawing/Processing/Extensions/ClearPathExtensions.cs:
```diff
@@ -1,79 +1,84 @@
-        }
+        /// <param name="source">The image processing context.</param>
+        /// <param name="color">The color.</param>
+        /// <param name="path">The logic path.</param>
+        /// <returns>The <see cref="IImageProcessingContext"/> to allow chaining of operations.</returns>
+        public static IImageProcessingContext Clear(
+            this IImageProcessingContext source,
+            Color color,
+            IPath path) =>
```
I would consider renaming this to `polygon`.

antonfirsov replied:
I would consider renaming these parameters to `region` so the name better represents the role of the parameter. Also, we definitely need more descriptive docs than "The logic path.".

```suggestion
        /// <param name="region">The <see cref="IPath"/> defining the region to fill.</param>
        /// <returns>The <see cref="IImageProcessingContext"/> to allow chaining of operations.</returns>
        public static IImageProcessingContext Clear(
            this IImageProcessingContext source,
            Color color,
            IPath region) =>
```

antonfirsov commented on src/ImageSharp.Drawing/Processing/Extensions/ClearPathExtensions.cs:
```diff
@@ -1,79 +1,84 @@
     {
         /// <summary>
-        /// Clones the shape graphic options and applies changes required to force clearing.
+        /// Flood fills the image in the path of the provided polygon with the specified brush without any blending.
```
```suggestion
        /// Flood fills the image within the provided region defined by an <see cref="IPath"/> using the specified brush without any blending.
```

antonfirsov commented on src/ImageSharp.Drawing/Processing/Processors/Drawing/FillPathProcessor.cs:
```diff
@@ -32,9 +36,9 @@ public FillPathProcessor(DrawingOptions options, IBrush brush, IPath shape)
 
         /// <summary>
-        /// Gets the region that this processor applies to.
+        /// Gets the logic path that this processor applies to.
```
```suggestion
        /// Gets the <see cref="Path"/> defining the region to fill.
```

antonfirsov commented on src/ImageSharp.Drawing/Processing/Processors/Drawing/FillPathProcessor.cs:
```diff
@@ -32,9 +36,9 @@ public FillPathProcessor(DrawingOptions options, IBrush brush, IPath shape)
+        /// Gets the logic path that this processor applies to.
         /// </summary>
-        public IPath Shape { get; }
+        public IPath Path { get; }
```
I think "region" is a better name when it's about filling:

```suggestion
        public IPath Region { get; }
```

antonfirsov commented on src/ImageSharp.Drawing/Processing/Processors/Drawing/RecursiveImageProcessor.cs:
```diff
@@ -0,0 +1,97 @@
+    /// <summary>
+    /// Allows the recursive application of processing operations against an image.
+    /// </summary>
+    public class RecursiveImageProcessor : IImageProcessor
```
`ProcessInsideRegionProcessor`?

antonfirsov commented on src/ImageSharp.Drawing/Processing/Extensions/ClipPathExtensions.cs:
```diff
@@ -0,0 +1,28 @@
+        /// <param name="path">The target path to operate within.</param>
+        /// <param name="operation">The operation to perform on the source.</param>
+        /// <returns>The <see cref="IImageProcessingContext"/> to allow chaining of operations.</returns>
+        public static IImageProcessingContext Clip(
```
I think `Clip` is a very generic overloaded and ambiguous name, doesn't help API discoverability in this case.

I wouldn'be afraid of going verbose here, for example `ProcessInsideRegion` would be a better name for this.

antonfirsov commented on src/ImageSharp.Drawing/Processing/Processors/Drawing/RecursiveImageProcessor.cs:
```diff
@@ -0,0 +1,97 @@
+        private class RecursiveImageProcessorInner<TPixel> : IImageProcessor<TPixel>
+            where TPixel : unmanaged, IPixel<TPixel>
+        {
+            private readonly RecursiveImageProcessor recursiveImageProcessor;
```
We are using the term `definition` or `processorDefinition` on other places. Describes the role between the non-generic and the generic thing better IMO.

antonfirsov commented on src/ImageSharp.Drawing/Processing/Processors/Drawing/RecursiveImageProcessor.cs:
```diff
@@ -0,0 +1,97 @@
+        /// <summary>
+        /// Gets the logic path.
+        /// </summary>
+        public IPath Path { get; }
```
```suggestion
        public IPath Region { get; }
```

antonfirsov reviewed (commented):
Implementation and tests look good.

We need to clean up the terminology used in API naming and docs, because currently it feels confusing to me. The problem is that `IPath` can be used for many things in the Drawing library, it's name doesn't carry the purpose (which I personally dislike, but it's a common practice in drawing libraries). My recommendation is to use the term "Region" (or an equivalent unambiguously descriptive term) for an `IPath` defining a region to fill, or boundary for a recursive processor.

We also need to find a better name for the new `Clip` method.

JimBobSquarePants commented:
@antonfirsov I've applied a lot of renaming and updated XML docs based on your suggestions. I found that I was straying into #94 territory though as I waded through each connected method so have stopped where I am for now. 

I maintained `Clip` as the method name (_I think people will get it_) and renamed the processor `ClipPathProcessor`

JimBobSquarePants commented:
I'm merging this. We can do a final API review before RC1
